adapter/internal/discovery/xds/semanticVersioning.go

```diff
@@ -0,0 +1,403 @@
+			apiVersionRegex := GetVersionMatchRegex(apiVersion)
+			if isLatestMajorVersion {
+				regex = strings.Replace(regex, apiVersionRegex, GetMajorMinorVersionRangeRegex(*apiSemVersion), 1)
+				regexRewritePattern = strings.Replace(regexRewritePattern, apiVersionRegex, GetMajorMinorVersionRangeRegex(*apiSemVersion), 1)
+			} else if isLatestMinorVersion {
+				regex = strings.Replace(regex, apiVersionRegex, GetMinorVersionRangeRegex(*apiSemVersion), 1)
+				regexRewritePattern = strings.Replace(regexRewritePattern, apiVersionRegex, GetMinorVersionRangeRegex(*apiSemVersion), 1)
+			}
+			pathSpecifier := &routev3.RouteMatch_SafeRegex{
+				SafeRegex: &envoy_type_matcherv3.RegexMatcher{
+					Regex: regex,
+				},
+			}
+			route.Match.PathSpecifier = pathSpecifier
+			action := &routev3.Route_Route{}
+			action = route.Action.(*routev3.Route_Route)
+			action.Route.RegexRewrite.Pattern.Regex = regexRewritePattern
+			route.Action = action
+		}
+	}
+}
+
+func updateRoutingRulesOnAPIDelete(organizationID, apiIdentifier string, api mgw.MgwSwagger) {
+	// Update the intelligent routing if the deleting API is the latest version of the API range
+	// and the API range has other versions
+	apiRangeIdentifier := GenerateIdentifierForAPIWithoutVersion(api.GetVHost(), api.GetTitle())
+
+	latestAPIVersionMap, latestAPIVersionMapExists := orgIDLatestAPIVersionMap[organizationID][apiRangeIdentifier]
+	if !latestAPIVersionMapExists {
+		return
+	}
+	deletingAPISemVersion, _ := ValidateAndGetVersionComponents(api.GetVersion(), api.GetTitle())
+	if deletingAPISemVersion == nil {
+		return
+	}
+	majorVersionRange := GetMajorVersionRange(*deletingAPISemVersion)
+	newLatestMajorRangeAPIUUID := ""
+	if deletingAPIsMajorRangeLatestAPISemVersion, ok := latestAPIVersionMap[majorVersionRange]; ok {
+		if deletingAPIsMajorRangeLatestAPISemVersion.Version == api.GetVersion() {
+			newLatestMajorRangeAPI := &SemVersion{
+				Version: "",
+				Major:   deletingAPISemVersion.Major,
+				Minor:   0,
+				Patch:   nil,
+			}
+			for apiUUID, swagger := range orgIDAPIMgwSwaggerMap[organizationID] {
+				// Iterate all the API versions other than the deleting API itself
+				if swagger.GetTitle() == api.GetTitle() && apiUUID != apiIdentifier {
+					currentAPISemVersion, _ := ValidateAndGetVersionComponents(swagger.GetVersion(), swagger.GetTitle())
+					if currentAPISemVersion != nil {
+						if currentAPISemVersion.Major == deletingAPISemVersion.Major {
+							if CompareSemanticVersions(*newLatestMajorRangeAPI, *currentAPISemVersion) {
+								newLatestMajorRangeAPI = currentAPISemVersion
+								newLatestMajorRangeAPIUUID = apiUUID
+							}
+						}
+					}
+				}
+			}
+			if newLatestMajorRangeAPIUUID != "" {
+				orgIDLatestAPIVersionMap[organizationID][apiRangeIdentifier][majorVersionRange] = *newLatestMajorRangeAPI
+				for _, route := range orgIDOpenAPIRoutesMap[organizationID][newLatestMajorRangeAPIUUID] {
```

Comment:
newLatestMajorRangeAPIUUID needs to be combination of vhost and uuid

Reply:
This is already a combination of vhost and uuid. Renamed the variable to avoid confusion